From 577f2402671d0fff0d9479ed3202a40f3b885b41 Mon Sep 17 00:00:00 2001
From: Nico Weber <thakis@chromium.org>
Date: Thu, 13 May 2021 16:07:29 +0000
Subject: [PATCH] build: Convert tools/metrics/BUILD.gn to py3

No changes to the python script are needed.

I checked that the generated histograms.xml is byte-for-byte identical
with python3 and python2.7. On my linux box, the script takes ~56s with
python2.7 and ~28s with python3, so this makes the step run twice as fast :)

While here, don't needlessly open() input files: Let minidom.parse()
do it for us. This is perf-neutral.

Also explicitly set the encoding and newline escaping on the output
file. This has no effect on my linux box, but makes it more likely
that the output histograms.xml is the same on all host systems.

Bug: 1205592
Change-Id: I22ff86552bd4e0aa14048583969b6400df7663fd
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2892629
Auto-Submit: Nico Weber <thakis@chromium.org>
Commit-Queue: Nico Weber <thakis@chromium.org>
Reviewed-by: Alexei Svitkine <asvitkine@chromium.org>
Cr-Commit-Position: refs/heads/master@{#882521}
---
 src/tools/metrics/BUILD.gn                | 5 +----
 src/tools/metrics/histograms/merge_xml.py | 5 +++--
 2 files changed, 4 insertions(+), 6 deletions(-)

diff --git a/src/tools/metrics/BUILD.gn b/src/tools/metrics/BUILD.gn
index b27f1096f6..ccb04c7295 100644
--- a/src/tools/metrics/BUILD.gn
+++ b/src/tools/metrics/BUILD.gn
@@ -2,10 +2,7 @@
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
 
-import("//build/config/python.gni")
-
-# TODO(crbug.com/1118214) - Make work under Python 3.
-python2_action("histograms_xml") {
+action("histograms_xml") {
   script = "histograms/merge_xml.py"
   output = "$root_out_dir/histograms.xml"
   outputs = [ output ]
diff --git a/src/tools/metrics/histograms/merge_xml.py b/src/tools/metrics/histograms/merge_xml.py
index d522fae6d1..a9ddd49613 100755
--- a/src/tools/metrics/histograms/merge_xml.py
+++ b/src/tools/metrics/histograms/merge_xml.py
@@ -182,7 +182,8 @@ def MergeFiles(filenames=[], files=[], should_expand_owners=False):
   Returns:
     A merged DOM tree.
   """
-  all_files = files + [open(f) for f in filenames]
+  # minidom.parse() takes both files and filenames:
+  all_files = files + filenames
   trees = [xml.dom.minidom.parse(f) for f in all_files]
   return MergeTrees(trees, should_expand_owners)
 
@@ -196,7 +197,7 @@ def main():
   parser = argparse.ArgumentParser()
   parser.add_argument('--output', required=True)
   args = parser.parse_args()
-  with open(args.output, 'w') as f:
+  with open(args.output, 'w', encoding='utf-8', newline='\n') as f:
     # This is run by
     # https://source.chromium.org/chromium/chromium/src/+/master:tools/metrics/BUILD.gn;drc=573e48309695102dec2da1e8f806c18c3200d414;l=5
     # to send the merged histograms.xml to the server side. Providing |UKM_XML|
